man/burst.man
.TH BURST %manext1% 2013-02-08 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
burst \- explode digests into nmh messages
.SH SYNOPSIS
.HP 5
.na
.B burst
.RB [ \-help ]
.RB [ \-version ]
.RI [ +folder ]
.RI [ msgs ]
.RB [ \-inplace " | " \-noinplace ]
.RB [ \-mime " | " \-nomime ]
.RB [ \-automime ]
.RB [ \-quiet " | " \-noquiet ]
.RB [ \-verbose " | " \-noverbose ]
.ad
.SH DESCRIPTION
.B burst
considers the specified messages in the named folder to be
Internet digests, and explodes them in that folder.
.PP
If
.B \-inplace
is given, each digest is replaced by its \*(lqtable
of contents\*(rq and the digest is removed.
.B burst
then renumbers the messages which follow the digest, in the folder,
making room for each of the exploded messages, which are placed
immediately after the \*(lqtable of contents\*(rq.
.PP
If
.B \-noinplace
is given, each digest is preserved, no table of contents
is produced, and the exploded messages are placed at the end of the
folder.  Other messages in the folder are not affected.
.PP
If
.B \-automime
is given,
.B burst
will try to determine if the message is formatted with MIME and contains MIME parts of
type \*(lqmessage/rfc822\*(rq.  If it does, it will burst the message using MIME
formatting rules.  The
.B \-mime
switch can be used to enforce the use of MIME formatting.  The
.B \-nomime
switch will force
.B burst
to use RFC 934 rules.
.PP
The
.B \-quiet
switch directs
.B burst
to be silent about reporting
messages that are not in digest format.
.PP
The
.B \-verbose
switch directs
.B burst
to tell the user the general
actions that it is taking to explode the digest.
.PP
It turns out that
.B burst
works equally well on forwarded messages
and blind-carbon-copies as on Internet digests, provided
that the former were generated by
.B forw
or
.BR send .
.SH FILES
.TP 20
$HOME/.mh_profile
The user's profile.
.SH "PROFILE COMPONENTS"
.PD 0
.TP 20
Path:
To determine the user's nmh directory.
.TP
Current\-Folder:
To find the default current folder.
.TP
Msg\-Protect:
To set mode when creating a new message.
.PD
.SH "SEE ALSO"
.IR inc (1),
.IR pack (1)
.PP
.I
Proposed Standard for Message Encapsulation
(RFC 934)
.SH DEFAULTS
.PD 0
.TP 20
+folder
The current folder.
.TP
msgs
The current message.
.TP
\-noinplace
.TP
\-automime
.TP
\-noquiet
.TP
\-noverbose
.PD
.SH CONTEXT
If a folder is given, it will become the current folder.  If
.B \-inplace
is given, then the first message burst becomes the current message.
This leaves the context ready for a
.B show
of the table of contents
of the digest, and a
.B next
to see the first message of the digest.  If
.B \-noinplace
is given, then the first message extracted from the
first digest burst becomes the current message.  This leaves the context
in a similar, but not identical, state to the context achieved when using
.BR \-inplace .
.SH BUGS
The
.B burst
program enforces a limit of approximately 1,000 on the number of
messages which may be
.B burst
from a single message. There is usually no limit on the number of
messages which may reside in the folder after the
.BR burst ing.
.PP
Although
.B burst
uses a sophisticated algorithm to determine where
one encapsulated message ends and another begins, not all digestifying
programs use an encapsulation algorithm.  In degenerate cases, this
usually results in
.B burst
finding an encapsulation boundary
prematurely and splitting a single encapsulated message into two or
more messages.  These erroneous digestifying programs should be fixed.
.PP
Furthermore, any text which appears after the last encapsulated message
is not placed in a separate message by
.BR burst .
In the case of
digestified messages, this text is usually an \*(lqEnd of digest\*(rq
string.  As a result of this possibly unfriendly behavior on the
part of
.BR burst ,
note that when the
.B \-inplace
option is used,
this trailing information is lost.  In practice, this is not a problem
since correspondents usually place remarks in text prior to the first
encapsulated message, and this information is not lost.


NEWS
#
# NEWS - Release notes for nmh 1.6
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been nearly two years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time
MH and nmh uses should read careful the NOTEABLE CHANGES section, as there
are some significant changes to nmh behavior.  Otherwise, please see the
README and INSTALL files for help on getting started with nmh.

------------
NEW FEATURES
------------

The following are new features for the 1.6 release of nmh:

- All nmh programs will now warn about multiple profile entries for
  the same non-null, non-comment component.  In previous versions, all
  but the first were silently ignored.
- mhmail now supports all post(8) options, or optionally can use
  send(1) with all of its options.  Its formerly undocumented -resent
  switch has been documented.  And it has new -attach, -headerfield,
  and -send/-nosend switches.
- Default to flock() locking on OpenBSD and Darwin.
- Added -directives support to mhbuild(1) to control whether or not
  mhbuild will honor MIME directives by default.  And added support
  for special #on/#off/#pop directives to control the MIME directive
  processing state.
- Added -messageid switch to send(1) and post(8).  This allows
  selection of the style to use for generated Message-ID and
  Content-ID header fields.  The default localname style is
  pid.time@localname, where time is in seconds, and matches previous
  behavior.  The random style replaces the localname with some
  (pseudo)random bytes and uses microsecond-resolution time.
- Added -clobber switch to mhstore(1) to control overwriting of
  existing files.
- Added -outfile switch to mhstore(1).
- Added -noall/-all switches to sortm(1).  sortm -noall requires
  a messages argument.
- $PAGER overrides the compiled-in default pager command.
- Added etc/scan.highlighted format file, as an example of how to
  highlight/colorize the output of scan(1).
- inc(1) now supports a -port switch to specify the port used by the
  POP server.
- pick(1) now decodes MIME-encoded header fields before searching.
- The VISUAL and EDITOR environment variables are now supported as fallbacks
  if the user does not configure an editor entry in their profile.
- The format engine (mh_format(5)) now properly accounts for multibyte
  characters when accounting for column widths.
- burst(1) now can burst MIME-formatted digests (messages that contain
  message/rfc822 parts instead of messages formatted with RFC 934).
- All proc entries (showproc, moreproc, etc) can now accept entries that
  contain spaces and shell metacharacters.  If found, such entries will
  either be space-split or processed by /bin/sh.
- A new program, fmttest(1) is included to help debug format files.  See
  NOTABLE CHANGES below.
- mhshow/mhstore now have support for RFC-2017 (access-type=url) for
  external message bodies.
- Added -retainsequences switch to refile(1).
- A new program, mhfixmsg(1), is included to rewrite MIME messages with
  various transformations.  See NOTABLE CHANGES below.
- Added -[no]rmmproc switches to rmm(1).
- Added support for Content-Disposition header (RFC 2183) to mhstore(1)
  and mhn(1) when used with -auto.
- All nmh commands now support transactional locking for sequence files.
  See NOTABLE CHANGES below.
- There is no longer a per-folder maximum number of sequences.
- For the SMTP MTA TLS can now be negotiated at the beginning of the
  connection with the -initialtls switch.
- Messages can now be selected using a relative offset from some other
  message, or from the start or end of a sequence.  See mh-sequence(5).
- The -changecur and -nochangecur switches have been added to mhlist(1).
- mhbuild(1) can now encode 8-bit message headers using RFC-2047 encoding
  rules.  See NOTABLE CHANGES below.
- The whatnow(1) attach feature will determine the content type of an
  attachment using a program such as file --mime-type, if available at
  configuration time.  If not, it will use mhshow-suffix- entries as
  before.  The -v switch to attach causes it to display the mhbuild
  directive that send(1) will use.
- mhbuild(1) now supports the -auto/-noauto flags (to be used by send(1)
  when invoking mhbuild automatically).
- mhbuild(1) now is automatically run by send, to insure that all outgoing
  messages have proper MIME formatting.  See NOTABLE CHANGES below.
- A new header, "Attach", is supported by mhbuild; it is used to replace
  previous functionality (which by default used a header named
  Nmh-Attachment).  See NOTABLE CHANGES below.
- The default Content-Transfer-Encoding for text parts is now 8bit.
- mhbuild(1) now supports a selectable Content-Transfer-Encoding
- If nmh was configured with iconv(3) support, mhshow will convert, if
  necessary, the charset of text/plain content to match the user's
  locale setting.  See NOTABLE CHANGES below.
- Added support for %{charset} display string escape to mhshow(1).
- The MIME parsing and generating routines now support RFC 2231 extended
  parameter information.
- mh-mime(7) now provides an introduction to nmh's MIME handling.
- mhshow(1) will now by default display all text content under one pager,
  and display markers for non-text and non-inline content.  The content
  markers are changeable via mh-format(5).  See NOTABLE CHANGES below.

-----------------
OBSOLETE FEATURES
-----------------
- Changed exit status of each nmh command's -version and -help
  switches from 1 to 0.
- The following environment variables were deprecated in nmh 1.5
  and removed from this release:
    MHPOPDEBUG (use -snoop command line switch instead)
    MM_NOASK (use -nolist and -nopause command line switches instead)
    NOMHNPROC (use -nocheckmime command line switch instead)
    FACEPROC (undocumented faceproc feature removed)
- Changed repl and dist default to -noatfile.  The default of -atfile

uip/burst.c
/* burst.c -- explode digests into individual messages
 *
 * This code is Copyright (c) 2002, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 */

#include <h/mh.h>
#include <h/utils.h>
#include <h/mhparse.h>

#define BURST_SWITCHES \
    X("inplace", 0, INPLSW) \
    X("noinplace", 0, NINPLSW) \
    X("mime", 0, MIMESW) \
    X("nomime", 0, NMIMESW) \
    X("automime", 0, AUTOMIMESW) \
    X("quiet", 0, QIETSW) \
    X("noquiet", 0, NQIETSW) \
    X("verbose", 0, VERBSW) \
    X("noverbose", 0, NVERBSW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(BURST);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(BURST, switches);
#undef X

struct smsg {
    off_t s_start;
    off_t s_stop;
};

/*
 * For the MIME parsing routines
 */

int debugsw = 0;

/*
 * static prototypes
 */
static int find_delim (int, struct smsg *, int *);
static void find_mime_parts (CT, struct smsg *, int *);
static void burst (struct msgs **, int, struct smsg *, int, int, int,
		   char *, int);
static void cpybrst (FILE *, FILE *, char *, char *, int, int);

/*
 * A macro to check to see if we have reached a message delimiter
 * (an encapsulation boundary, EB, in RFC 934 parlance).
 *
 * According to RFC 934, an EB is simply a line which starts with
 * a "-" and is NOT followed by a space.  So even a single "-" on a line
 * by itself would be an EB.
 */

#define CHECKDELIM(buffer) (buffer[0] == '-' && buffer[1] != ' ')

int
main (int argc, char **argv)
{
    int inplace = 0, quietsw = 0, verbosw = 0, mimesw = 1;
    int hi, msgnum, numburst;
    char *cp, *maildir, *folder = NULL, buf[BUFSIZ];
    char **argp, **arguments;
    struct msgs_array msgs = { 0, 0, NULL };
    struct smsg *smsgs;
    struct msgs *mp;

    if (nmh_init(argv[0], 1)) { return 1; }

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
	    case AMBIGSW: 
		ambigsw (cp, switches);
		done (1);
	    case UNKWNSW: 
		adios (NULL, "-%s unknown\n", cp);

	    case HELPSW: 
		snprintf (buf, sizeof(buf), "%s [+folder] [msgs] [switches]",
			invo_name);
		print_help (buf, switches, 1);
		done (0);
	    case VERSIONSW:
		print_version(invo_name);
		done (0);

	    case INPLSW: 
		inplace++;
		continue;
	    case NINPLSW: 
		inplace = 0;
		continue;

	    case MIMESW:
	    	mimesw = 2;
		continue;
	    case NMIMESW:
	    	mimesw = 0;
		continue;
	    case AUTOMIMESW:
	    	mimesw = 1;
		continue;

	    case QIETSW: 
		quietsw++;
		continue;
	    case NQIETSW: 
		quietsw = 0;
		continue;

	    case VERBSW: 
		verbosw++;
		continue;
	    case NVERBSW: 
		verbosw = 0;
		continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (folder)
		adios (NULL, "only one folder at a time!");
	    else
		folder = pluspath (cp);
	} else {
	    app_msgarg(&msgs, cp);
	}
    }

    if (!context_find ("path"))
	free (path ("./", TFOLDER));
    if (!msgs.size)
	app_msgarg(&msgs, "cur");
    if (!folder)
	folder = getfolder (1);
    maildir = m_maildir (folder);

    if (chdir (maildir) == NOTOK)
	adios (maildir, "unable to change directory to");

    /* read folder and create message structure */
    if (!(mp = folder_read (folder, 1)))
	adios (NULL, "unable to read folder %s", folder);

    /* check for empty folder */
    if (mp->nummsg == 0)
	adios (NULL, "no messages in %s", folder);

    /* parse all the message ranges/sequences and set SELECTED */
    for (msgnum = 0; msgnum < msgs.size; msgnum++)
	if (!m_convert (mp, msgs.msgs[msgnum]))
	    done (1);
    seq_setprev (mp);	/* set the previous-sequence */

    smsgs = mh_xcalloc(MAXFOLDER + 2, sizeof *smsgs);

    hi = mp->hghmsg + 1;

    /* burst all the SELECTED messages */
    for (msgnum = mp->lowsel; msgnum <= mp->hghsel; msgnum++) {
	if (is_selected (mp, msgnum)) {
	    if ((numburst = find_delim (msgnum, smsgs, &mimesw)) >= 1) {
		if (verbosw)
		    printf ("%d message%s exploded from digest %d\n",
			    numburst, numburst > 1 ? "s" : "", msgnum);
		burst (&mp, msgnum, smsgs, numburst, inplace, verbosw,
		       maildir, mimesw);
	    } else {
		if (numburst == 0) {
		    if (!quietsw)
			inform("message %d not in digest format, continuing...",
				  msgnum);
		}  /* this pair of braces was missing before 1999-07-15 */
		else
		    adios (NULL, "burst() botch -- you lose big");
	    }
	}
    }

    free(smsgs);
    context_replace (pfolder, folder);	/* update current folder */

    /*
     * If -inplace is given, then the first message burst becomes
     * the current message (which will now show a table of contents).
     * Otherwise, the first message extracted from the first digest
     * becomes the current message.
     */
    if (inplace) {
	if (mp->lowsel != mp->curmsg)
	    seq_setcur (mp, mp->lowsel);
    } else {
	if (hi <= mp->hghmsg)
	    seq_setcur (mp, hi);
    }

    seq_save (mp);	/* synchronize message sequences */
    context_save ();	/* save the context file         */
    folder_free (mp);	/* free folder/message structure */
    done (0);
    return 1;
}


/*
 * Scan the message and find the beginning and
 * end of all the messages in the digest.
 *
 * If requested, see if the message is MIME-formatted and contains any
 * message/rfc822 parts; if so, burst those parts.
 */

static int
find_delim (int msgnum, struct smsg *smsgs, int *mimesw)
{
    int wasdlm = 0, msgp;
    off_t pos;
    char c, *msgnam;
    char buffer[BUFSIZ];
    FILE *in;
    CT content;

    msgnam = m_name (msgnum);

    /*
     * If mimesw is 1 or 2, try to see if it's got proper MIME formatting.
     */

    if (*mimesw > 0) {
    	content = parse_mime(msgnam);
	if (! content && *mimesw == 2)
	    return 0;
	if (content) {
	    smsgs[0].s_start = 0;
	    smsgs[0].s_stop = content->c_begin - 1;
	    msgp = 1;
	    find_mime_parts(content, smsgs, &msgp);
	    free_content(content);
	    if (msgp == 1 && *mimesw == 2) {
	    	adios (msgnam, "does not have any message/rfc822 parts");
	    } else if (msgp > 1) {
	    	*mimesw = 1;
		return (msgp - 1);
	    }
	}
    }

    *mimesw = 0;

    if ((in = fopen (msgnam, "r")) == NULL)
	adios (msgnam, "unable to read message");

    for (msgp = 0, pos = 0L; msgp <= MAXFOLDER;) {
    	/*
	 * We're either at the beginning of the whole message, or
	 * we're just past the delimiter of the last message.
	 * Swallow lines until we get to something that's not a newline
	 */
	while (fgets (buffer, sizeof(buffer), in) && buffer[0] == '\n')
	    pos += (long) strlen (buffer);
	if (feof (in))
	    break;

	/*
	 * Reset to the beginning of the last non-blank line, and save our
	 * starting position.  This is where the encapsulated message
	 * starts.
	 */
	fseeko (in, pos, SEEK_SET);
	smsgs[msgp].s_start = pos;

	/*
	 * Read in lines until we get to a message delimiter.
	 *
	 * Previously we checked to make sure the preceding line and
	 * next line was a newline.  That actually does not comply with
	 * RFC 934, so make sure we break on a message delimiter even
	 * if the previous character was NOT a newline.
	 */
	for (c = 0; fgets (buffer, sizeof(buffer), in); c = buffer[0]) {
	    if ((wasdlm = CHECKDELIM(buffer)))
		break;
	    else
		pos += (long) strlen (buffer);
	}

	/*
	 * Only count as a new message if we got the message delimiter.
	 * Swallow a blank line if it was right before the message delimiter.
	 */
	if (smsgs[msgp].s_start != pos && wasdlm)
	    smsgs[msgp++].s_stop = (c == '\n' && wasdlm) ? pos - 1 : pos;

	if (feof (in)) {
#if 0
	    if (wasdlm) {
		smsgs[msgp - 1].s_stop -= ((long) strlen (buffer) + 1);
		msgp++;		/* fake "End of XXX Digest" */
	    }
#endif
	    break;
	}
	pos += (long) strlen (buffer);
    }

    fclose (in);
    return (msgp - 1);		/* return the number of messages burst */
}


/*
 * Find any MIME content in the message that is a message/rfc822 and add
 * it to the list of messages to burst.
 */

static void
find_mime_parts (CT content, struct smsg *smsgs, int *msgp)
{
    struct multipart *m;
    struct part *part;

    /*
     * If we have a message/rfc822, then it's easy.
     */

    if (content->c_type == CT_MESSAGE &&
    			content->c_subtype == MESSAGE_RFC822) {
	smsgs[*msgp].s_start = content->c_begin;
	smsgs[*msgp].s_stop = content->c_end;
	(*msgp)++;
	return;
    }

    /*
     * Otherwise, if we do have multiparts, try all of the sub-parts.
     */

    if (content->c_type == CT_MULTIPART) {
    	m = (struct multipart *) content->c_ctparams;

	for (part = m->mp_parts; part; part = part->mp_next)
	    find_mime_parts(part->mp_part, smsgs, msgp);
    }
}


/*
 * Burst out the messages in the digest into the folder
 */

static void
burst (struct msgs **mpp, int msgnum, struct smsg *smsgs, int numburst,
	int inplace, int verbosw, char *maildir, int mimesw)
{
    int i, j, mode;
    char *msgnam;
    char f1[BUFSIZ], f2[BUFSIZ], f3[BUFSIZ];
    FILE *in, *out;
    struct stat st;
    struct msgs *mp;

    if ((in = fopen (msgnam = m_name (msgnum), "r")) == NULL)
	adios (msgnam, "unable to read message");

    mode =
      fstat (fileno(in), &st) != NOTOK ? (int) (st.st_mode & 0777) : m_gmprot();
    mp = *mpp;

    /*
     * See if we have enough space in the folder
     * structure for all the new messages.
     */
    if ((mp->hghmsg + numburst > mp->hghoff) &&
	!(mp = folder_realloc (mp, mp->lowoff, mp->hghmsg + numburst)))
	adios (NULL, "unable to allocate folder storage");
    *mpp = mp;

    j = mp->hghmsg;		/* old value */
    mp->hghmsg += numburst;
    mp->nummsg += numburst;

    /*
     * If this is not the highest SELECTED message, then
     * increment mp->hghsel by numburst, since the highest
     * SELECTED is about to be slid down by that amount.
     */
    if (msgnum < mp->hghsel)
	mp->hghsel += numburst;

    /*
     * If -inplace is given, renumber the messages after the
     * source message, to make room for each of the messages
     * contained within the digest.
     *
     * This is equivalent to refiling a message from the point
     * of view of the external hooks.
     */
    if (inplace) {
	for (i = mp->hghmsg; j > msgnum; i--, j--) {
	    strncpy (f1, m_name (i), sizeof(f1));
	    strncpy (f2, m_name (j), sizeof(f2));
	    if (does_exist (mp, j)) {
		if (verbosw)
		    printf ("message %d becomes message %d\n", j, i);

		if (rename (f2, f1) == NOTOK)
		    admonish (f1, "unable to rename %s to", f2);

		(void)snprintf(f1, sizeof (f1), "%s/%d", maildir, i);
		(void)snprintf(f2, sizeof (f2), "%s/%d", maildir, j);
		ext_hook("ref-hook", f1, f2);

		copy_msg_flags (mp, i, j);
		clear_msg_flags (mp, j);
		mp->msgflags |= SEQMOD;
	    }
	}
    }
    
    unset_selected (mp, msgnum);

    /* new hghmsg is hghmsg + numburst
     *
     * At this point, there is an array of numburst smsgs, each element of
     * which contains the starting and stopping offsets (seeks) of the message
     * in the digest.  The inplace flag is set if the original digest is replaced
     * by a message containing the table of contents.  smsgs[0] is that table of
     * contents.  Go through the message numbers in reverse order (high to low).
     *
     * Set f1 to the name of the destination message, f2 to the name of a scratch
     * file.  Extract a message from the digest to the scratch file.  Move the
     * original message to a backup file if the destination message number is the
     * same as the number of the original message, which only happens if the
     * inplace flag is set.  Then move the scratch file to the destination message.
     *
     * Moving the original message to the backup file is equivalent to deleting the
     * message from the point of view of the external hooks.  And bursting each
     * message is equivalent to adding a new message.
     */

    i = inplace ? msgnum + numburst : mp->hghmsg;
    for (j = numburst; j >= (inplace ? 0 : 1); i--, j--) {
        char *tempfile;

	if ((tempfile = m_mktemp2(NULL, invo_name, NULL, &out)) == NULL) {
	    adios(NULL, "unable to create temporary file in %s",
		  get_temp_dir());
	}
	strncpy (f2, tempfile, sizeof(f2));
	strncpy (f1, m_name (i), sizeof(f1));

	if (verbosw && i != msgnum)
	    printf ("message %d of digest %d becomes message %d\n", j, msgnum, i);

	chmod (f2, mode);
	fseeko (in, smsgs[j].s_start, SEEK_SET);
	cpybrst (in, out, msgnam, f2,
		(int) (smsgs[j].s_stop - smsgs[j].s_start), mimesw);
	fclose (out);

	if (i == msgnum) {
	    strncpy (f3, m_backup (f1), sizeof(f3));
	    if (rename (f1, f3) == NOTOK)
		admonish (f3, "unable to rename %s to", f1);

	    (void)snprintf(f3, sizeof (f3), "%s/%d", maildir, i);
	    ext_hook("del-hook", f3, NULL);
	}
	if (rename (f2, f1) == NOTOK)
	    admonish (f1, "unable to rename %s to", f2);

	(void)snprintf(f3, sizeof (f3), "%s/%d", maildir, i);
	ext_hook("add-hook", f3, NULL);

	copy_msg_flags (mp, i, msgnum);
	mp->msgflags |= SEQMOD;
    }

    fclose (in);
}


#define	S1  0
#define	S2  1
#define	S3  2
#define S4  3

/*
 * Copy a message which is being burst out of a digest.
 * It will remove any "dashstuffing" in the message.
 */

static void
cpybrst (FILE *in, FILE *out, char *ifile, char *ofile, int len, int mime)
{
    int c, state;

    for (state = mime ? S4 : S1; (c = fgetc (in)) != EOF && len > 0; len--) {
	if (c == 0)
	    continue;
	switch (state) {
	    case S1: 
		switch (c) {
		    case '-': 
			state = S3;
			break;

		    default: 
			state = S2;
			/* FALLTHRU */
		    case '\n': 
			fputc (c, out);
			break;
		}
		break;

	    case S2: 
		switch (c) {
		    case '\n': 
			state = S1;
			/* FALLTHRU */
		    default: 
			fputc (c, out);
			break;
		}
		break;

	    case S3: 
		switch (c) {
		    case ' ': 
			state = S2;
			break;

		    default: 
			state = (c == '\n') ? S1 : S2;
			fputc ('-', out);
			fputc (c, out);
			break;
		}
		break;

	    case S4:
	   	fputc (c, out);
		break;
	}
    }

    if (ferror (in) && !feof (in))
	adios (ifile, "error reading");
    if (ferror (out))
	adios (ofile, "error writing");
}
